### PR TITLE
Adding a cache layer for avoiding to hit the storage - Saving network roundtrips

### PR DESCRIPTION
**Pull request by andrehjr**
High level config:

```
FeatureFlagger.configure do |config|
  redis = ::Redis::Namespace.new('feature-flagger-rollout-control', redis: RedisFactory.default_client)
  config.storage = FeatureFlagger::Storage::Redis.new(redis))
  config.cache_store = :memory_store, { expires_in: 10.minutes }
end
```

Cache Store is powered by ActiveSupport. So it works pretty much like https://guides.rubyonrails.org/caching_with_rails.html#cache-stores

The same options will apply. :file_store, :memory_store, etc

### CODE REVIEW

File: spec/feature_flagger/model_spec.rb [hunk 1]
```diff
@@ -44,9 +44,14 @@ def id; 14 end
...
 
         it 'calls Control#released? with appropriated methods' do
-          expect(control).to receive(:released?).with(resolved_key, resource_id)
+          expect(control).to receive(:released?).with(resolved_key, resource_id, {})
           DummyClass.released_id?(resource_id, key)
         end
+
+        it 'passes down cache options to storage' do
+          expect(control).to receive(:released?).with(resolved_key, resource_id, { skip_cache_store: true })
+          DummyClass.released_id?(resource_id, key, skip_cache_store: true)
```
Comment (henrich-m) [inline spec/feature_flagger/model_spec.rb:53]:
```suggestion
          DummyClass.released_id?(resource_id, key, skip_cache: true)
```

File: spec/feature_flagger/control_spec.rb [hunk 2]
```diff
@@ -137,6 +166,26 @@ module FeatureFlagger
...
+
+      context 'when caching is configured' do
+        let(:cache_store) { ActiveSupport::Cache::MemoryStore.new }
+
+        it 'only hits the storage once' do
+          expect(storage).to receive(:all_values).once
+          2.times { control.resource_ids(key) }
+        end
+
+        it 'hits the storage n times when skip_cache_store is provided' do
```
Comment (henrich-m) [inline spec/feature_flagger/control_spec.rb:183]:
```suggestion
        it 'hits the storage n times when skip_cache param is provided' do
```

File: spec/feature_flagger/model_spec.rb [hunk 3]
```diff
@@ -44,9 +44,14 @@ def id; 14 end
         let(:resource_id) { 10 }
 
         it 'calls Control#released? with appropriated methods' do
-          expect(control).to receive(:released?).with(resolved_key, resource_id)
+          expect(control).to receive(:released?).with(resolved_key, resource_id, {})
           DummyClass.released_id?(resource_id, key)
         end
+
+        it 'passes down cache options to storage' do
+          expect(control).to receive(:released?).with(resolved_key, resource_id, { skip_cache_store: true })
```
Comment (henrich-m) [inline spec/feature_flagger/model_spec.rb:52]:
```suggestion
          expect(control).to receive(:released?).with(resolved_key, resource_id, { skip_cache: true })
```

File: spec/feature_flagger/model_spec.rb [hunk 4]
```diff
@@ -74,9 +79,14 @@ def id; 14 end
 
     describe '.all_released_ids_for' do
       it 'calls Control#resource_ids with appropriated methods' do
-        expect(control).to receive(:resource_ids).with(resolved_key)
+        expect(control).to receive(:resource_ids).with(resolved_key, {})
         DummyClass.all_released_ids_for(key)
       end
+
+      it 'passes down cache options to storage' do
+        expect(control).to receive(:resource_ids).with(resolved_key, { skip_cache_store: true})
```
Comment (henrich-m) [inline spec/feature_flagger/model_spec.rb:87]:
```suggestion
        expect(control).to receive(:resource_ids).with(resolved_key, { skip_cache: true})
```

File: spec/feature_flagger/model_spec.rb [hunk 5]
```diff
@@ -74,9 +79,14 @@ def id; 14 end
...
+        DummyClass.all_released_ids_for(key, skip_cache_store: true)
```
Comment (henrich-m) [inline spec/feature_flagger/model_spec.rb:88]:
```suggestion
        DummyClass.all_released_ids_for(key, skip_cache: true)
```

**Review by nandosousafr — APPROVED** [review]
:trophy: 

Thank you @andrehjr

**Pull request merged**